### PR TITLE
Allow users to withdraw their tokens on L1

### PR DESCRIPTION
**Pull request by alexkeating**
#### Description

- Handles withdrawing tokens on L1 by burning tokens on L2 and sending a message.
TODO: add deploy scripts and test with the whole system

### CODE REVIEW

**Comment by alexkeating** [comment]
Current dependencies on/for this PR:
* main
  * **PR #11** <a href="https://app.graphite.dev/github/pr/ScopeLift/l2-flexible-voting/11" target="_blank">Graphite</a> 
    * **PR #15** <a href="https://app.graphite.dev/github/pr/ScopeLift/l2-flexible-voting/15" target="_blank">Graphite</a> 
      * **PR #8** <a href="https://app.graphite.dev/github/pr/ScopeLift/l2-flexible-voting/8" target="_blank">Graphite</a> 
        * **PR #9** <a href="https://app.graphite.dev/github/pr/ScopeLift/l2-flexible-voting/9" target="_blank">Graphite</a> 
          * **PR #10** <a href="https://app.graphite.dev/github/pr/ScopeLift/l2-flexible-voting/10" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ScopeLift/l2-flexible-voting/10?utm_source=stack-comment).

File: src/L2ERC20.sol [hunk 1]
```diff
@@ -44,4 +47,15 @@ contract L2ERC20 is ERC20Votes, WormholeReceiver {
     require(clock() == L1_BLOCK.number(), "ERC20Votes: broken clock mode");
     return "mode=blocknumber&from=eip155:1";
   }
+
+  /// @notice Burn L2 tokens and unlock tokens on the L1.
+  /// @param account The account where the tokens will be transferred.
+  /// @param amount The amount of tokens to be unlocked.
+  function l1Unlock(address account, uint256 amount) external returns (uint256 sequence) {
+    _burn(account, amount);
```
Comment (apbendi) [inline src/L2ERC20.sol:55]: I think this should be `msg.sender` otherwise anyone can call this for any L2 holder 😬

File: src/L1ERC20Bridge.sol [hunk 2]
```diff
@@ -50,4 +52,19 @@ contract L1ERC20Bridge is L1VotePool {
     nonce = nonce + 1;
     return sequence;
   }
+
+  /// @notice Receives an encoded withdrawal message from the L2
+  /// @param encodedMsg The encoded message from the L2 with the withdrawal information.
+  function receiveEncodedWithdrawalMsg(bytes memory encodedMsg) public {
```
Comment (apbendi) [inline src/L1ERC20Bridge.sol:58]: This function can now receive two types of messages:

1. Vote messages
2. Withdrawal messages

How does Wormhole know which function signature to call with each type of bridge message? Should we also change the name of the vote receiving method to make it clear what it's for?

**Review by apbendi — CHANGES_REQUESTED** [review]
Some initial feedback @alexkeating

File: src/L1ERC20Bridge.sol (hunk 2 above)
Comment (alexkeating) [reply src/L1ERC20Bridge.sol:58]: With the specialized relayer we would write the relayer call. So, the method name could be anything as long as we coded the specialized relayer properly. In a future pr when we switch to the generic relayer the method name is sent to the generic relayer.

Comment (alexkeating) [reply src/L1ERC20Bridge.sol:58]: We should definitely update the name to make it more clear.

File: src/L2ERC20.sol (hunk 1 above)
Comment (alexkeating) [reply src/L2ERC20.sol:56]: https://github.com/ScopeLift/l2-flexible-voting/pull/10/commits/ad70b0db9ad45595bd5f879b0e5489dde17a9ce4

File: src/L1ERC20Bridge.sol (hunk 2 above)
Comment (alexkeating) [reply src/L1ERC20Bridge.sol:58]: So, `receiveEncodedMsg` is in the `WormholeReceiver` interface. Renaming should be handled in the generic relayer pr.

File: src/L2ERC20.sol (hunk 1 above)
Comment (apbendi) [reply src/L2ERC20.sol:56]: I don't think the change is what we want. Correct me if I'm wrong but my understanding is:

`account` is the address where the tokens will be sent on L1
`msg.sender` is the L2 token holder that is bridging them back to L1

Often they'll be the same, but not always. So it's ok for `account` to be something other than `msg.sender`. But the `_burn` call should use `msg.sender`, while the message passed via the bridge should use `account`

Comment (alexkeating) [reply src/L2ERC20.sol:56]: Ah, I see now, yes that makes sense

Comment (alexkeating) [reply src/L2ERC20.sol]: https://github.com/ScopeLift/l2-flexible-voting/pull/10/commits/546c9f6d77ad36f09f1fa87790c21b73756c7628

**Review by apbendi — APPROVED** [review]

**Pull request merged**